src/vm/crossloaderallocatorhash.inl

```diff
@@ -0,0 +1,1196 @@
+
+    bool result = true;
+
+    GCPROTECT_BEGIN(gc);
+    {
+        gc.tracker->GetDependentAndLoaderAllocator(&gc.loaderAllocatorRef, &gc.keyToValuesHashObject);
+        if (gc.keyToValuesHashObject != NULL)
+        {
+            gc.keyToValuesHash = KeyToValuesGCHeapHash(gc.keyToValuesHashObject);
+            INT32 indexInKeyValueHash = gc.keyToValuesHash.GetValueIndex(&key);
+            if (indexInKeyValueHash != -1)
+            {
+                gc.keyToValuesHash.GetElement(indexInKeyValueHash, gc.keyValueStore);
+
+                result = VisitKeyValueStore(&gc.loaderAllocatorRef, &gc.keyValueStore, visitor);
+            }
+        }
+    }
+    GCPROTECT_END();
+
+    return result;
+}
+
+template <class TRAITS>
+template <class Visitor>
+/*static*/ bool CrossLoaderAllocatorHash<TRAITS>::VisitTrackerAllEntries(LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe, Visitor &visitor)
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    struct _gcStruct
+    {
+        LAHASHDEPENDENTHASHTRACKERREF tracker;
+        OBJECTREF loaderAllocatorRef;
+        GCHEAPHASHOBJECTREF keyToValuesHashObject;
+        KeyToValuesGCHeapHash keyToValuesHash;
+        OBJECTREF keyValueStore;
+    }gc;
+
+    class VisitAllEntryKeyValueHash
+    {
+        public:
+        Visitor *_pVisitor;
+        KeyToValuesGCHeapHash *_pKeysToValueHash;
+        OBJECTREF *_pKeyValueStore;
+        OBJECTREF *_pLoaderAllocatorRef;
+
+        VisitAllEntryKeyValueHash(Visitor *pVisitor,  KeyToValuesGCHeapHash *pKeysToValueHash, OBJECTREF *pKeyValueStore, OBJECTREF *pLoaderAllocatorRef) : 
+            _pVisitor(pVisitor),
+            _pKeysToValueHash(pKeysToValueHash),
+            _pKeyValueStore(pKeyValueStore),
+            _pLoaderAllocatorRef(pLoaderAllocatorRef)
+            {}
+
+        bool operator()(INT32 index)
+        {
+            WRAPPER_NO_CONTRACT;
+
+            _pKeysToValueHash->GetElement(index, *_pKeyValueStore);
+            return VisitKeyValueStore(_pLoaderAllocatorRef, _pKeyValueStore, visitor);
+        }
+    };
+
+    ZeroMemory(&gc, sizeof(gc));
+    gc.tracker = trackerUnsafe;
+
+    bool result = true;
+
+    GCPROTECT_BEGIN(gc);
+    {
+        gc.tracker->GetDependentAndLoaderAllocator(&gc.loaderAllocatorRef, &gc.keyToValuesHashObject);
+        if (gc.keyToValuesHashObject != NULL)
+        {
+            gc.keyToValuesHash = KeyToValuesGCHeapHash(gc.keyToValuesHashObject);
+            result = gc.keyToValuesHash.VisitAllEntryIndices(VisitAllEntryKeyValueHash(&visitor, &gc.keyToValuesHash, &gc.keyValueStore, &gc.loaderAllocatorRef));
+        }
+    }
+    GCPROTECT_END();
+
+    return result;
+}
+
+template <class TRAITS>
+template <class Visitor>
+/*static*/ bool CrossLoaderAllocatorHash<TRAITS>::VisitKeyToTrackerAllEntries(OBJECTREF hashKeyEntryUnsafe, Visitor &visitor)
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    struct _gcStruct
```

Comment:
A nit - can you please remove the _gcStruct? All other places in the code use anonymous structs for the GC protection purpose

Reply:
Removed